src/compiler/checker.ts

```diff
@@ -393,9 +398,28 @@ namespace ts {
             }
 
             let sourceFiles = host.getSourceFiles();
-            return sourceFiles.indexOf(file1) <= sourceFiles.indexOf(file2);
-        }
+            return sourceFiles.indexOf(file1) <= sourceFiles.indexOf(file2) || !isInSameLexicalScope(); 
+
+            function isInSameLexicalScope() {
+                let node1Scope = getEnclosingBlockScopeContainer(node1);
+                let node2Scope = getEnclosingBlockScopeContainer(node2);
+                
+                while (node2Scope && node1Scope !== node2Scope) {
+                    switch (node2Scope.kind) {
+                        // In function like scope
+                        case SyntaxKind.FunctionExpression:
+                        case SyntaxKind.FunctionDeclaration:
+                        case SyntaxKind.ArrowFunction:
+                            return false;
+                    }
 
+                    node2Scope = getEnclosingBlockScopeContainer(node2Scope);

```

Comment:
We should never run into these kind of situation. I am going to modify parameter names to take declaration and reference so it is more clear